jans-lock/cedarling/src/lock_master/mod.rs

```diff
@@ -0,0 +1,109 @@
+use std::borrow::Cow;
+use wasm_bindgen::{prelude::*, throw_val};
+use web_sys::*;
+
+use crate::{
+	http::{self, ResponseEx},
+	startup::types::PolicyStoreConfig,
+};
+
+pub(crate) mod types;
+
+#[wasm_bindgen]
+extern "C" {
+	#[wasm_bindgen(js_name = btoa)]
+	pub fn js_btoa(input: &str) -> String;
+}
+
+pub(crate) async fn init<'a>(policy_store_config: &PolicyStoreConfig) -> Cow<'a, [u8]> {
+	let PolicyStoreConfig::LockMaster {
+		url,
+		policy_store_id,
+		enable_dynamic_configuration,
+		ssa_jwt,
+		application_name,
+	} = policy_store_config
+	else {
+		unreachable!("We arrive here from the PolicyStoreConfig::LockMaster")
+	};
+
+	// Get LockMasterConfig
+	let res = http::get(&format!("{}/.well-known/lock-master-configuration", url), &[])
+		.await
+		.expect_throw("Unable to fetch LockMasterConfig from URL");
+	let lock_master_config: types::LockMasterConfig = res.into_json::<types::LockMasterConfig>().await.unwrap_throw();
+
+	// Setup dynamic SSE updates for LockMaster
+	if *enable_dynamic_configuration {
+		let sse = EventSource::new(&lock_master_config.lock_sse_uri).unwrap_throw();
+		let sse2 = sse.clone();
+
+		let onopen = Closure::once_into_js(move || {
+			let onmessage = Closure::<dyn Fn(MessageEvent)>::new(move |ev: MessageEvent| {
+				console::log_2(&JsValue::from_str("Cedarling Received message: "), &ev);
+				unimplemented!("Dynamic Configuration Updates")
+			})
+			.into_js_value();
+			sse2.set_onmessage(Some(onmessage.unchecked_ref()));
+
+			let onerror = Closure::<dyn Fn(JsValue)>::new(move |ev: JsValue| throw_val(ev)).into_js_value();
+			sse2.set_onerror(Some(onerror.unchecked_ref()));
+		});
+
+		sse.set_onopen(Some(onopen.unchecked_ref()));
+	}
+
+	// Get OAuthConfig
+	let res = http::get(&lock_master_config.oauth_as_well_known, &[]).await.expect_throw("Unable to fetch LockMasterConfig from URL");
+	let openid_config = res.into_json::<types::OAuthConfig>().await.unwrap_throw();
+
+	let client: types::OAuthDynamicClient = {
+		// OpenID dynamic client registration
+		let client_req = types::OAuthDynamicClientRequest {
+			client_name: application_name,
+			application_type: "web",
+			redirect_uris: &[],
+			token_endpoint_auth_method: "client_secret_basic",
+			software_statement: ssa_jwt,
+			contacts: &["newton@gluu.org"],
+		};
+
+		// send
+		let res = http::post(&openid_config.registration_endpoint, http::PostBody::Json(client_req), &[])
+			.await
+			.expect_throw("Unable to register client");
+		res.into_json().await.expect_throw("Unable to parse client registration response")
+	};
+
+	let grant: types::OAuthGrantResponse = {
+		// https://gluu.org/docs/gluu-server/4.0/admin-guide/oauth2/
+		let token = if let Some(client_secret) = client.client_secret {
+			format!("{}:{}", client.client_id, client_secret)
+		} else {
+			console::warn_1(&JsValue::from_str("Client Secret is not provided"));
```

Comment:
I'm not sure why this check is needed. You always get a client ID and secret back from DCR.

Reply:
I actually don't remember what frame of reference I got this from, but it _can_ miss in some circumstances. Besides, it's a check for that edge case and doesn't affect normal behaviour

Reply:
Sure, but leaving possibly dead code will make it harder to audit later on. Please annotate with a todo specifying the edge case and if this can be removed in the future.